plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpoolingManager.java

```diff
@@ -0,0 +1,179 @@
+import java.util.Optional;
+
+import static io.airlift.slice.Slices.utf8Slice;
+import static io.trino.spooling.filesystem.EncryptionUtils.decryptingInputStream;
+import static io.trino.spooling.filesystem.EncryptionUtils.encryptingOutputStream;
+import static io.trino.spooling.filesystem.EncryptionUtils.generateRandomKey;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class FileSystemSpoolingManager
+        implements SpoolingManager
+{
+    private static final Logger log = Logger.get(FileSystemSpoolingManager.class);
+
+    private final String location;
+    private final TrinoFileSystem fileSystem;
+    private final Duration ttl;
+    private final boolean encryptionEnabled;
+
+    @Inject
+    public FileSystemSpoolingManager(FileSystemSpoolingConfig config, TrinoFileSystemFactory fileSystemFactory)
+    {
+        requireNonNull(config, "config is null");
+        this.location = config.getLocation();
+        this.fileSystem = requireNonNull(fileSystemFactory, "fileSystemFactory is null")
+                .create(ConnectorIdentity.ofUser("ignored"));
+        this.ttl = config.getTtl();
+        this.encryptionEnabled = config.isEncryptionEnabled();
+    }
+
+    @Override
+    public OutputStream createOutputStream(SpooledSegmentHandle handle)
+            throws Exception
+    {
+        FileSystemSpooledSegmentHandle filesystemHandle = (FileSystemSpooledSegmentHandle) handle;
+        OutputStream stream = fileSystem.newOutputFile(segmentLocation(filesystemHandle)).create();
+        if (filesystemHandle.encryptionKey().isPresent()) {
+            return encryptingOutputStream(stream, filesystemHandle.encryptionKey().get());
+        }
+        return stream;
+    }
+
+    @Override
+    public FileSystemSpooledSegmentHandle create(SpoolingContext context)
+    {
+        if (encryptionEnabled) {
+            return FileSystemSpooledSegmentHandle.random(context, ttl, Optional.of(generateRandomKey()));
+        }
+        return FileSystemSpooledSegmentHandle.random(context, ttl);
+    }
+
+    @Override
+    public InputStream openInputStream(SpooledSegmentHandle handle)
+            throws IOException
+    {
+        FileSystemSpooledSegmentHandle segmentHandle = (FileSystemSpooledSegmentHandle) handle;
+        checkExpiration(segmentHandle);
+        try {
+            if (!fileSystem.newInputFile(segmentLocation(segmentHandle)).exists()) {
+                throw new RuntimeException("Segment not found or expired");
+            }
+
+            InputStream stream = fileSystem.newInputFile(segmentLocation((FileSystemSpooledSegmentHandle) handle)).newStream();
+            return new ExceptionMappingInputStream(segmentHandle.encryptionKey().map(key -> {
+                try {
+                    return decryptingInputStream(stream, key);
+                }
+                catch (IOException e) {
+                    throw new RuntimeException(e);
```

Comment:
We shouldn't rethrow here, as `openInputStream()` is expected to throw `IOException`